Mathlib/CategoryTheory/Dialectica/Basic.lean

```diff
@@ -0,0 +1,98 @@
+/-
+Copyright (c) 2024 Mario Carneiro. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Mario Carneiro
+-/
+import Mathlib.CategoryTheory.Subobject.Basic
+import Mathlib.CategoryTheory.Limits.Shapes.FiniteProducts
+
+/-!
+# Dialectica category
+
+We define the category `Dial` of the Dialectica interpretation, after:
+
+* Valeria de Paiva, The Dialectica Categories.
+  University of Cambridge, Computer Laboratory, PhD Thesis, Technical Report 213, 1991
+  ([pdf](https://www.cl.cam.ac.uk/techreports/UCAM-CL-TR-213.pdf)).
+-/
+
+noncomputable section
+
+namespace CategoryTheory
+
+open Limits
+
+theorem Functor.map_le_map {C D} [Preorder C] [Preorder D]
+    (F : C ⥤ D) {X Y : C} (h : X ≤ Y) : F.obj X ≤ F.obj Y := (F.map h.hom).le
+
+universe v u
+variable {C : Type u} [Category.{v} C] [HasFiniteProducts C] [HasPullbacks C]
+
+variable (C) in
+/-- The Dialectica category. An object of the category is a triple `⟨U, X, α ⊆ U × X⟩`,
+and a morphism from `⟨U, X, α⟩` to `⟨V, Y, β⟩` is a pair `(f : U ⟶ V, F : U ⨯ Y ⟶ X)` such that
+`α(u, F(u, y)) ≤ β(f(u), y)`. -/
+structure Dial where
+  /-- The source object -/
+  src : C
+  /-- The target object -/
+  tgt : C
+  /-- A subobject of `src ⨯ tgt`, interpreted as a relation -/
+  rel : Subobject (src ⨯ tgt)
+
+namespace Dial
+
+local notation "π₁" => prod.fst
+local notation "π₂" => prod.snd
+local notation "π(" a ", " b ")" => prod.lift a b
+
+/-- A morphism in the `Dial C` category from `⟨U, X, α⟩` to `⟨V, Y, β⟩` is a pair
+`(f : U ⟶ V, F : U ⨯ Y ⟶ X)` such that `α(u, F(u, y)) ≤ β(f(u), y)`. -/
```

Comment:
The expression `α(u, F(u, y)) ≤ β(f(u), y)` does not fully makes sense to me. At the very least, I find it more confusing. It would be nice to at least say explicitly that this is an inequality in subobjects of `V ⨯ Y`. Also, writing `{ α(u, F(u, y)) } ≤ { β(f(u), y) }` would makes things more clear.

Reply:
Would it be better to use subset notation like `{ (u,y) | α(u, F(u, y)) } ⊆{ (u,y) | β(f(u), y) }`? For me the definition with point-free descriptions of pullbacks is completely impenetrable, and I need to frequently consult this "internal language" version in order to make sense of things. (It's also not an inequality in `V ⨯ Y`, it's in `U ⨯ Y`. I was hoping the choice of letters would make the typing clear.)